File: Manh/bai-tap-buoi-so-4.hs
{-# LANGUAGE OverloadedStrings #-}
module Example where

import Language.Marlowe.Extended.V1

main :: IO ()
main = printJSON example


{- Define a contract, Close is the simplest contract which just ends the contract straight away
-}

-- Các bên tham gia vào hợp đồng
chuDoanhNghiep = Role "ChuDoanhNghiep"
nhanVien1 = Role "NhanVien1"
nhanVien2 = Role "NhanVien2"
nhanVien3 = Role "NhanVien3"
nhanVien4 = Role "NhanVien4"
nhanVien5 = Role "NhanVien5"
nhanVien6 = Role "NhanVien6"
nhanVien7 = Role "NhanVien7"
nhanVien8 = Role "NhanVien8"
nhanVien9 = Role "NhanVien9"
nhanVien10 = Role "NhanVien10"

--Các tham số dùng trong hợp đồng
thoiHanChuyenLuong = TimeParam "ThoiHanChuyenLuong"
luong10NhanVien = ConstantParam "Luong10NhanVien"
luong1NhanVien = DivValue luong10NhanVien (Constant 10)

-- hàm này sẽ thực hiện lệnh Pay để trả lương cho 1 nhân viên từ tài khoản của chủ doanh nghiệp
traLuong nhanVien buocTiepTheo = 
    Pay
        chuDoanhNghiep
        (Party nhanVien)
        ada
        luong1NhanVien
        buocTiepTheo 

chuyenTienLuongVaoHopDong buocTiepTheo = 
    When
        [Case
            (Deposit
                chuDoanhNghiep
                chuDoanhNghiep
                ada
                luong10NhanVien
            )
            buocTiepTheo]
        (TimeParam "ThoiHanChuyenLuong")
        Close 

example :: Contract
example = chuyenTienLuongVaoHopDong (
            traLuong nhanVien1 (
                traLuong nhanVien2 (
                    traLuong nhanVien3 (
                        traLuong nhanVien4 (
                            traLuong nhanVien5 (
                                traLuong nhanVien6 (
                                    traLuong nhanVien7 (
                                        traLuong nhanVien8 (
                                            traLuong nhanVien9 (
                                                traLuong nhanVien10 Close
                                            )
                                        )
                                    )
                                )
                            )
                        )
                    )
                )
            )
        )

File: Manh/bai-tap-buoi-so-4.ts
import {
    Address, Role, Account, Party, ada, lovelace, AvailableMoney, Constant, ConstantParam,
    NegValue, AddValue, SubValue, MulValue, DivValue, ChoiceValue, TimeIntervalStart,
    TimeIntervalEnd, UseValue, Cond, AndObs, OrObs, NotObs, ChoseSomething,
    ValueGE, ValueGT, ValueLT, ValueLE, ValueEQ, TrueObs, FalseObs, Deposit,
    Choice, Notify, Close, Pay, If, When, Let, Assert, SomeNumber, AccountId,
    ChoiceId, Token, ValueId, Value, EValue, Observation, Bound, Action, Payee,
    Case, Timeout, ETimeout, TimeParam, Contract
} from 'marlowe-js';

(function (): Contract {

    // Các bên tham gia vào hợp đồng
    const chuDoanhNghiep : Party = Role("ChuDoanhNghiep");
    const nhanVien1 : Party = Role("NhanVien1");
    const nhanVien2 : Party = Role("NhanVien2");
    const nhanVien3 : Party = Role("NhanVien3");
    const nhanVien4 : Party = Role("NhanVien4");
    const nhanVien5 : Party = Role("NhanVien5");
    const nhanVien6 : Party = Role("NhanVien6");
    const nhanVien7 : Party = Role("NhanVien7");
    const nhanVien8 : Party = Role("NhanVien8");
    const nhanVien9 : Party = Role("NhanVien9");
    const nhanVien10 : Party = Role("NhanVien10");

    // Các tham số dùng trong hợp đồng
    const thoiHanChuyenLuong : Timeout = TimeParam("ThoiHanChuyenLuong");
    const luong10NhanVien : Value = ConstantParam("Luong10NhanVien");
    const luong1NhanVien : Value = DivValue(luong10NhanVien, Constant(10n));

    function traLuong (nhanVien: Party, buocTiepTheo: Contract) {
        return Pay(chuDoanhNghiep, Party(nhanVien), lovelace, luong1NhanVien, buocTiepTheo);
    }

    function chuyenTienLuongVaoHopDong (buocTiepTheo: Contract) {
        return When([
            Case(Deposit(chuDoanhNghiep, chuDoanhNghiep, lovelace, luong10NhanVien), buocTiepTheo)
        ], thoiHanChuyenLuong, Close);
    }

    return chuyenTienLuongVaoHopDong (
        traLuong (nhanVien1, 
            traLuong (nhanVien2, 
                traLuong (nhanVien3, 
                    traLuong (nhanVien4, 
                        traLuong (nhanVien5, 
                            traLuong (nhanVien6, 
                                traLuong (nhanVien7, 
                                    traLuong (nhanVien8, 
                                        traLuong (nhanVien9, 
                                            traLuong (nhanVien10, Close)
                                        )
                                    )
                                )
                            )
                        )
                    )
                )
            )
        )
    );
})
